Fix incorrect Reinhard tonemap operator

tracefree opened the pull request:
Fixes #93317 

This PR changes the formula for applying the Reinhard tonemap operator to correctly use equation 4 in [Reinhard's orginal 2002 paper](https://www-old.cs.utah.edu/docs/techreports/2002/pdf/UUCS-02-001.pdf) and updates relevant documentation accordingly. For more details see the discussion in #93317.

Here are some screenshots of the TPS demo with and without the fix applied, taken at various whitepoint values:

### With the fix
`white = 6.0`. This is the intended setting for the scene.
![With the fix at white = 6.0](https://github.com/godotengine/godot/assets/155747722/52325b29-9bd0-4457-9db8-895c0a7c80f9)
`white = 1.0`
![With the fix at white = 1.0](https://github.com/godotengine/godot/assets/155747722/5f4c991e-3e25-43c1-afd4-2edcb294c228)
`white = 0.5`
![With the fix at white = 0.5](https://github.com/godotengine/godot/assets/155747722/192d67f7-0504-4d39-ada2-cdef1e4de410)

### Without the fix (current master branch)
`white = 6.0`
![Without the fix at white = 6.0](https://github.com/godotengine/godot/assets/155747722/a43d8f66-3b04-43b8-8f80-726b698aac07)
`white = 1.0`
![Without the fix at white = 1.0](https://github.com/godotengine/godot/assets/155747722/61d118ea-c859-47be-9097-7ec8e11e2030)
`white = 0.5`
![Without the fix at white = 0.5](https://github.com/godotengine/godot/assets/155747722/b6f3b570-440e-4d94-b4be-82ec4dbe43d7)

tracefree committed:
Fix incorrect Reinhard tonemap operator

AThousandShips commented on doc/translations/zh_CN.po:
Don't edit the translations, please restore these

tracefree replied:
Done, thanks for letting me know!

clayjohn commented on servers/rendering/renderer_rd/shaders/effects/tonemap.glsl:
```diff
@@ -244,7 +244,8 @@ vec3 tonemap_aces(vec3 color, float white) {
 vec3 tonemap_reinhard(vec3 color, float white) {
-	return (white * color + color) / (color * white + white);
+	float white_squared = white * white;
+	return (white_squared * color + color * color) / (color * white_squared + white_squared);
```
```suggestion
	return (white_squared * color + color) / (color * white_squared + white_squared);
```

Looks like an extra `color` slipped in here.

tracefree replied:
The extra color is part of the fix. It is needed to conform to the formula given in Reinhard's paper. That it was missing in Godot's implementation is the main issue.

tracefree replied:
Fixed the implementation for the compatibility renderer as well - thanks, I don't know how I missed that!

clayjohn replied:
I haven't looked at the paper, but from your issue report, you say that Reinhard's formula is:
```
color * (1 + 1/(white * white)) / (color + 1)
```
While we use:
```
(white * color + color) / (color * white + white);
```
If you divide the entire thing by `white` you get:
```
(color + (color/white)) / (color + 1)
```
And if you factor out `color` from the top half you get Reinhard's original formula (with white being broken of course):
```
color * (1 + 1/(white)) / (color + 1)
```

This new version you have added turns into:
```
color * (1 + color / (white * white)) / (color + 1)
```

tracefree replied:
Ah, I'm sorry, I made a mistake in my issue report! The formula (4) in Reinhard's paper is
```
color * (1 + color/(white * white)) / (color + 1)
```
![Screenshot_20240618_235824](https://github.com/godotengine/godot/assets/155747722/d2ad9cfc-29f6-48ee-8866-1edd7e0d02e4)
L(x,y) is our `color`, and L_white is `white`. So my fix is correct, I just made a typo in the issue, gonna fix that right away.

allenwp replied:
PR as it stands [appears to be correct](https://www.desmos.com/calculator/2wcdxmi8dq).

I presume that it's better performance to be written like this, otherwise matching the original formula's format would be a touch easier to read if someone is comparing it to the paper in the future.

clayjohn replied:
Ah, that makes sense. So this is a bigger change than we thought. 

> I presume that it's better performance to be written like this, otherwise matching the original formula's format would be a touch easier to read if someone is comparing it to the paper in the future.

Ya, performance is better the way it was originally written. 

**OLD**
This is 3 instructions (1 divide, 2 FMAs)
```
(white * color + color) / (color * white + white);
```

This is 5 instructions (2 divides, 2 additions, 1 multiply):
```
color * (1 + 1/(white)) / (color + 1)
```

**New**
This is 5 instructions (2 multiply, 2 FMA, 1 divide)
```
float white_squared = white * white;
(white_squared * color + (color * color)) / (color * white_squared + white_squared);
```

This is 6 instructions (2 multiply, 2 addition, 2 divide)
```
color * (1 + color / (white * white)) / (color + 1)
```

tracefree replied:
I added comments to the code showing the non-simplified formula and a link to the paper, so in the future people can more easily check the source.

Chaosus replied:
```suggestion
	float white_squared_color = white_squared * color;
	return (white_squared_color  + color * color) / (white_squared_color  + white_squared);
```

`white_squared * color` and `color * white_squared` are identical? - I suggest pushing it to float constant and use it in both places.

Chaosus replied:
```suggestion
	return fma(white_squared, color, color * color) / fma(color, white_squared, white_squared);
```

If the suggestion above is not desired, consider to use built-in `fma` function.

tracefree replied:
Good catch, thank you!

clayjohn reviewed (commented):
You will also need to change the code for the Compatibility renderer here: https://github.com/godotengine/godot/blob/e2fc6d38cb240ea23bd5700514f8a8ce7011e89c/drivers/gles3/shaders/tonemap_inc.glsl#L80

tracefree commented:
I fail a Godot CPP test because I changed the enum name TONE_MAPPER_REINHARDT to TONE_MAPPER_REINHARD. Should I just revert that change? It's not essential to the bug fix, I just thought I'd fix the spelling while I'm at it, but I realize now that this would break compatibility with Godot CPP.

clayjohn commented:
> I fail a Godot CPP test because I changed the enum name TONE_MAPPER_REINHARDT to TONE_MAPPER_REINHARD. Should I just revert that change? It's not essential to the bug fix, I just thought I'd fix the spelling while I'm at it, but I realize now that this would break compatibility with Godot CPP.

Ya, you should revert it. It sucks to have a misspelled enum name. But it is more important to maintain compatibility then it is to correct a spelling error

tracefree commented:
I reverted the enum name change, added comments linking to the paper in the code, and made the use of whitespace in the forumula given in the documentation consistent.

allenwp commented:
Here's another comparison of before and after using some [test scenes I just whipped up](https://github.com/allenwp/calinou-godot-demo-projects/tree/add-color-correction-demo/3d/tonemap_color_correction):

_All images use Forward+ with whitepoint 6.0_
| Tonemap | Gradients (Linear Scale, No Debanding) | Hues (Exponential Scale, No Debanding)
|---|---|---
| Reinhard | ![Gradients-Reinhard](https://github.com/user-attachments/assets/cb7b1718-feb0-4a05-bebf-efd1c292bef8) | ![Hues-Reinhard](https://github.com/user-attachments/assets/a5c21f1d-043b-48e2-81a2-43834f08964a)
| Corrected Reinhard PR #93324 | ![Gradients-Corrected-Reinhard](https://github.com/user-attachments/assets/e69b5f1f-47a1-4ac6-b749-31289f0c9aa2) | ![Hues-Corrected-Reinhard](https://github.com/user-attachments/assets/1a1c0057-e761-43b1-93d4-e844b1ea54dc)

clayjohn reviewed (approved):
Sorry, I lost track of this after my last round of reviews. It looks good to me! 

Just for posterity, I commented on the increase in instructions in a previous review. It makes sense for us to use this optimized version over the naive formula. But I am not concerned about the overall increase in instructions from using the correct formula for two reasons:
1. Its the correct formula, using an incorrect formula isn't a good optimization
2. This shader is texture read bound (especially if using glow, or FXAA etc.), so the extra instructions will be hidden by the usual latency hiding mechanisms. 

It still makes sense to use the optimized form for two reasons:
1. Its better for mobile devices as it will use less battery (despite not running faster)
2. Its better if we end up copying this code in another part of the engine that may be more instruction-sensitive

clayjohn commented:
Just since it has stagnated for awhile, do you mind doing an empty force push just to force the CI to run again?

akien-mga commented:
I pushed a rebase for good measure as the base branch for this PR was 2500 commits behind `master`.

akien-mga commented:
Thanks! And congrats for your first merged Godot contribution :tada:

tracefree commented:
Thank you, and thanks to everyone who helped! :D

MJacred commented:
@akien-mga: I believe it makes sense to cherrypick this PR?

allenwp commented:
Although the old Reinhard implementation was clearly incorrect, it wasn’t necessarily problematic for games that used it. After this fix, lighting or colour adjustments may need to be re-tweaked for some games, so this kind of fix shouldn’t come through in a minor release; in my opinion it should only be introduced in a major release, like 4.4.0, when users are expecting these sorts of changes.
